### PR TITLE
Added test scene with attach_screen

### PR DESCRIPTION
**Pull request by clayjohn**
The proposed optimization [here](https://github.com/godotengine/godot/issues/26440) relies on using ``VisualServer.viewport_attach_to_screen`` to render viewports directly to the screen instead of through a ``ViewportContainer`` (or in your case a ``TextureRect``). The optimization would avoid having to copy the screen multiple times as your ``Viewport`` is transferred to the root-Viewport then the root-Viewport is transferred to the screen. 

This PR establishes a sample case that uses ``attach_to_screen`` instead of ``TextureRects`` so that I could get a good baseline for how much the optimization actually improves rendering. As it turns out, **without the optimization**, just using ``attach_to_screen`` decreases rendering time dramatically. 

### Performance Comparison
I tested on a chromebook with an Intel(R) Celeron(R) CPU  N2840  @ 2.16GHz processor and integrated graphics in GLES3 and Godot version 3.1.

On "Main.tscn" I was fluctuating between about 30-40ms for render time (~25-35 fps). This is due to the sheer size of the render target involved. 

Performance increased dramatically using "Node.tscn" **without optimization**. In "Node.tscn" I got a stable 15-20ms render time (50-60 fps). Again, the sheer size of the render target makes these numbers unsurprising. 

### Explanation.
In "Main.tscn" you render the entire thing to a giant (1920, 2160) viewport and then slice it up using texture rects. Under the hood the process is:
1. Render to Viewport ->
2. Read from Viewport, render to root->
3. copy from root to screen.

In "Node.tscn" the process is:
1. Render to Viewport 1 + Render to Viewport 2 ->
2. copy Viewport 1 to screen + Copy Viewport 2 to screen.

Root is never used so you skip an entire step. Each step represents a significant change in GPU state (which is the slowest part of rendering), also since Viewport 1 and 2 combined are equal in size to Viewport rendering both takes about the same time as just rendering the one. 

Without the optimization, using ``attach_to_screen`` already provides huge performance gains. 

#### The Optimization
The proposed optimization will even further increase render times by turning the process in Node. tscn into:
1. Render Viewport 1 to screen + Render Viewport 2 to screen.

Please let me know if you would like a detailed explanation about how "Node.tscn" works, I would be happy to explain the process.

### CODE REVIEW

**Comment by Keetz** [comment]
In Main.tscn I have ~1330 fps and in Node.tscn I have ~1370 fps.
Not sure where I can see the render time?

Two  problems I encounter using _viewport_attach_to_screen_:
1. When resizing the window, everything disappears.
2. I can't seem to easily move the targets. For example in Main.tscn I just set the position on the TextureRect if I want the side by side instead.

I have a few questions to #2.
Why do you halve the size here:
```
#First viewport will take up the top half of the screen
$Viewport.set_attach_to_screen_rect(Rect2(0, 0, 960, 540))
#Second viewport takes up the bottom half of the screen
$Viewport2.set_attach_to_screen_rect(Rect2(0, 540, 960, 540))
```
My test use case is:
Go to project settings and set the window width to **3840** and height to **1080**.
Set the two halves of the game side by side instead of on top of eachother.

What came first to me was to change the position in _set_attach_to_screen_rect()_ so it looks like this:
```
#First viewport will take up the top half of the screen
$Viewport.set_attach_to_screen_rect(Rect2(0, 0, 960, 540))
#Second viewport takes up the bottom half of the screen
$Viewport2.set_attach_to_screen_rect(Rect2(960, 0, 960, 540))
```
This is the result:
Selection_061

And this is the expected result:
Selection_062

**Comment by clayjohn** [comment]
> In Main.tscn I have ~1330 fps and in Node.tscn I have ~1370 fps.
> Not sure where I can see the render time?
The extra screen copy makes much less of a difference on good hardware. This is why I tested on a low-end cpu with only integrated graphics. Render time is just 1/FPS.

> When resizing the window, everything disappears.
I can't seem to easily move the targets. For example in Main.tscn I just set the position on the TextureRect if I want the side by side instead.

Control nodes like TextureRect are set to automatically resize when moved, or if the window is resized. If you are going to use this optimization then we would have to set those sorts of things up manually. E.g. when the window_resized signal is fired, we would have to manually set the sizes and positions of the viewports, including re-shrinking the root viewport.

> I have a few questions to #2.
> Why do you halve the size here:

In the project settings the test_size was set to (960, 1080), So I set the viewports to take up half of that size. In the final version you will want to dynamically set the sizes to half of what your screen size is. I don't have a screen that big, so it would not be beneficial to set the screen so large anyway. 

The numbers used right now are magic numbers to make the viewports fit on the screen. This setup is just for testing and gaining some numbers for the optimization. It is by no means a production ready setup.

**Comment by Keetz** [comment]
I am not sure I understand the difference here, I guess it is how **Viewport.set_attach_to_screen_rect()** works that is confusing me.

If I understand how Godot does scaling of viewport correct I don't think the fact that Control nodes scale automatically matters.

The root viewport has a size that is equal to the window size (without border and top bar), but it also has a size_override which is always the (1920, 2160) set by the project settings in this case, no matter the window size.

Not 100% correct because the Window.Stretch options has an effect on how the viewport behaves, but you get the idea.

When the window  is resized the viewport size follows but Godot does some kind of scaling behind our back, because the size_override stays (1920, 2160) and ALL nodes has their original size. The TextureRects for example will always have the size (1920, 2160) with an y scale at 0.5.

I know control knows will follow the viewport if anchors are used, but none are in this case so it isn't because of control nodes that they follow if I understand this correct. Node2D's follow the viewport too because of this "magic" scaling Godot does of a viewport and everything inside.

Basically I don't really understand why your improvements work differently, is it because of the **Viewport.set_attach_to_screen_rect()**? Or is it because of the use of cameras?

**Comment by clayjohn** [comment]
> Also, I still don't understand why everything disappears on resizing the window. I could understand if the sizes just wouldn't fit the window correct anymore, or if stuff moved around.

The first line in "Node.gd" resizes the root Viewport to 0. This is done because the root Viewport is drawn last, so it ends up being drawn on top of all the others. When the window is resized, the root Viewport is also resized and ends up drawing over the entire window. To take advantage of this optimization you would have to poll every frame and force the root Viewport to size 0 again on a window resize. 

> Basically I don't really understand why your improvements work differently, is it because of how Viewport.set_attach_to_screen_rect() works? Or is it because of the use of cameras?

It is because of the way ``Viewport.set_attach_to_screen_rect()`` works. It bypasses a lot of the normal work that godot does and renders the Viewport directly to the screen. Maybe it will help if I went into more detail about the normal rendering process. Everything you have in your above comment is correct, but we bypass it by using attach_to_screen_rect.

In Godot everything is rendered to a root Viewport, so when you have a running scene your structure looks like this

Root Viewport
-> Node A
-> Node B
-> Viewport A
 ----> Node C
First, node C is rendered to Viewport A. Then, nodes A and B are rendered to the root VIewport. Finally, the root viewport is copied over to the screen. 

If we want to see the contents of Viewport A, we can attach its texture to either Node A or B (or in your case A _and_ B). And the process becomes, node C is rendered to Viewport A. Then, nodes A and B grab Viewport A's texture and draw it to the root viewport. Finally the root Viewport is copied to the screen.

In low-end devices performing all these copies can be slow. At the very least they add unnecessary overhead. So, instead of using nodes A and B to render Viewport A to the root Viewport, we skip over that and copy Viewport A directly to the screen. This bypasses the root Viewport and nodes A and B. However, as you pointed out above, Godot handles scaling (especially GUI) using control nodes and the root Viewport. By using ``set_attach_to_screen_rect`` we are losing some of the benefits that control nodes provide. 

So, back to your original question. By bypassing the root Viewport, you have to manually control the size that Viewport A takes up on the screen. You can no longer trust Godot to do that for you. And you have to specify coordinates in window coordinates as you are rendering directly to the screen.

Because we are copying the full viewport directly to the screen, we can no longer chop it up. Before, we could use a texture with a shader that only read from one part of Viewport A's texture, making it easy to slice it up any direction. Now, we have to use two Viewports, each with a dedicated camera, the camera determines which area of the game scene gets drawn. So that only the parts we want are rendered to each Viewport and we don't have to perform the slicing after the fact.

**Comment by Keetz** [comment]
Makes sense. Thanks for clarifying!
